### ISSUE TITLE
Fix codecov reports

### ISSUE THREAD

**Issue by grigoryk** [mozilla-mobile/android-components#9382]
Couldn't find an exact issue for this, so let's just start here.

Currently even though codecov is supposedly enabled, we don't get any reports. Looking at the logs from a recent PR push, I see the following:
```
[task 2021-01-12T17:08:32.623Z] + automation/taskcluster/action/upload_coverage_report.sh
[task 2021-01-12T17:08:32.623Z] ++ export CI_BUILD_URL=https://tools.taskcluster.net/tasks/cRTJUbTqS5el82Ws0lLf5A
[task 2021-01-12T17:08:32.623Z] ++ CI_BUILD_URL=https://tools.taskcluster.net/tasks/cRTJUbTqS5el82Ws0lLf5A
[task 2021-01-12T17:08:32.623Z] ++ bash automation/taskcluster/.codecov -t @.cc_token
[task 2021-01-12T17:08:32.628Z] 
[task 2021-01-12T17:08:32.628Z]   _____          _
[task 2021-01-12T17:08:32.628Z]  / ____|        | |
[task 2021-01-12T17:08:32.628Z] | |     ___   __| | ___  ___ _____   __
[task 2021-01-12T17:08:32.628Z] | |    / _ \ / _` |/ _ \/ __/ _ \ \ / /
[task 2021-01-12T17:08:32.628Z] | |___| (_) | (_| |  __/ (_| (_) \ V /
[task 2021-01-12T17:08:32.628Z]  \_____\___/ \__,_|\___|\___\___/ \_/
[task 2021-01-12T17:08:32.628Z]                               Bash-20200602-f809a24
[task 2021-01-12T17:08:32.628Z] 
[task 2021-01-12T17:08:32.628Z] 
[task 2021-01-12T17:08:32.629Z] x> No CI provider detected.
[task 2021-01-12T17:08:32.629Z]     Testing inside Docker? http://docs.codecov.io/docs/testing-with-docker
[task 2021-01-12T17:08:32.629Z]     Testing with Tox? https://docs.codecov.io/docs/python#section-testing-with-tox
[task 2021-01-12T17:08:32.629Z]     project root: .
[task 2021-01-12T17:08:32.642Z]     Yaml found at: codecov.yml
[task 2021-01-12T17:08:32.669Z] ==> Running gcov in . (disable via -X gcov)
[task 2021-01-12T17:08:32.802Z] ==> Python coveragepy not found
[task 2021-01-12T17:08:32.802Z] ==> Searching for coverage reports in:
[task 2021-01-12T17:08:32.802Z]     + .
[task 2021-01-12T17:08:33.335Z] --> No coverage report found.
[task 2021-01-12T17:08:33.335Z]     Please visit http://docs.codecov.io/docs/supported-languages
```

A good place to start seems like this line: `Python coveragepy not found`

**Comment by grigoryk** [comment]
Prior issue: https://github.com/mozilla-mobile/android-components/issues/7500 (uploads timing out)

**Comment by grigoryk** [comment]
From .codecov:
```
if which coverage >/dev/null 2>&1;
 ....good stuff...
else
 say "${e}==>${x} Python coveragepy disabled"
fi
```

**Comment by grigoryk** [comment]
Comparable job running in Fenix (coverage reports work there):
```
[task 2021-01-12T14:42:57.322Z] + automation/taskcluster/upload_coverage_report.sh
[task 2021-01-12T14:42:57.323Z] ++ export CI_BUILD_URL=https://tools.taskcluster.net/tasks/T5pK_CzqRn-dEdWgKt8f4A
[task 2021-01-12T14:42:57.323Z] ++ CI_BUILD_URL=https://tools.taskcluster.net/tasks/T5pK_CzqRn-dEdWgKt8f4A
[task 2021-01-12T14:42:57.323Z] ++ bash automation/taskcluster/.codecov -t @.cc_token
[task 2021-01-12T14:42:57.330Z] 
[task 2021-01-12T14:42:57.330Z]   _____          _
[task 2021-01-12T14:42:57.330Z]  / ____|        | |
[task 2021-01-12T14:42:57.330Z] | |     ___   __| | ___  ___ _____   __
[task 2021-01-12T14:42:57.330Z] | |    / _ \ / _` |/ _ \/ __/ _ \ \ / /
[task 2021-01-12T14:42:57.330Z] | |___| (_) | (_| |  __/ (_| (_) \ V /
[task 2021-01-12T14:42:57.330Z]  \_____\___/ \__,_|\___|\___\___/ \_/
[task 2021-01-12T14:42:57.330Z]                               Bash-20200602-f809a24
[task 2021-01-12T14:42:57.330Z] 
[task 2021-01-12T14:42:57.330Z] 
[task 2021-01-12T14:42:57.331Z] x> No CI provider detected.
[task 2021-01-12T14:42:57.331Z]     Testing inside Docker? http://docs.codecov.io/docs/testing-with-docker
[task 2021-01-12T14:42:57.331Z]     Testing with Tox? https://docs.codecov.io/docs/python#section-testing-with-tox
[task 2021-01-12T14:42:57.332Z]     project root: .
[task 2021-01-12T14:42:57.347Z]     Yaml found at: codecov.yml
[task 2021-01-12T14:42:57.379Z] ==> Running gcov in . (disable via -X gcov)
[task 2021-01-12T14:42:57.449Z] ==> Python coveragepy not found
[task 2021-01-12T14:42:57.449Z] ==> Searching for coverage reports in:
[task 2021-01-12T14:42:57.449Z]     + .
[task 2021-01-12T14:42:57.775Z]     -> Found 1 reports
[task 2021-01-12T14:42:57.775Z] ==> Detecting git/mercurial file structure
[task 2021-01-12T14:42:57.788Z] ==> Reading reports
[task 2021-01-12T14:42:57.792Z]     + ./app/build/reports/jacoco/jacocoDebugTestReport/jacocoDebugTestReport.xml bytes=5880938
[task 2021-01-12T14:42:57.799Z] ==> Appending adjustments
[task 2021-01-12T14:42:57.799Z]     https://docs.codecov.io/docs/fixing-reports
[task 2021-01-12T14:42:59.557Z]     + Found adjustments
[task 2021-01-12T14:42:59.559Z] ==> Gzipping contents
[task 2021-01-12T14:42:59.882Z] ==> Uploading reports
[task 2021-01-12T14:42:59.882Z]     url: https://codecov.io
[task 2021-01-12T14:42:59.882Z]     query: branch=&commit=23a95450621c0d1cfce56180cc3a632a82f65322&build=&build_url=https://tools.taskcluster.net/tasks/T5pK_CzqRn-dEdWgKt8f4A&name=&tag=&slug=mozilla-mobile%2Ffenix&service=&flags=&pr=&job=
[task 2021-01-12T14:42:59.885Z]     -> Pinging Codecov
[task 2021-01-12T14:42:59.885Z] https://codecov.io/upload/v4?package=bash-20200602-f809a24&token=secret&branch=&commit=23a95450621c0d1cfce56180cc3a632a82f65322&build=&build_url=https://tools.taskcluster.net/tasks/T5pK_CzqRn-dEdWgKt8f4A&name=&tag=&slug=mozilla-mobile%2Ffenix&service=&flags=&pr=&job=
[task 2021-01-12T14:43:00.227Z]     -> Uploading
[task 2021-01-12T14:43:00.469Z]     -> View reports at https://codecov.io/github/mozilla-mobile/fenix/commit/23a95450621c0d1cfce56180cc3a632a82f65322
```

`coveragepy` is still not found, but a report is being generated nonetheless. So that's a red herring.

**Comment by grigoryk** [comment]
`jacocoTestReport` is registered as a `finalizedBy` task if `coverage` flag is set. All that seems to be happening, and the task does appear in the task graph list, but it's being SKIPPED.

We can see that in the CI logs:
`Task :feature-syncedtabs:jacocoTestReport SKIPPED`

**Comment by grigoryk** [comment]
Locally I've changed how `executionData` is specified to use `getExecutionData().setFrom(...)` form, and since we use `includeNoLocationClasses` had to specify `jacoco.excludes = ['jdk.internal.*']` to get tests to run locally with a `-Pcoverage` flag.

```
getExecutionData().setFrom(fileTree(dir: project.buildDir, includes: [
                            'jacoco/testDebugUnitTest.exec', 'outputs/code-coverage/connected/*coverage.ec'
                    ]))
```

After making these changes, the I see the following after running tests:

```
➜  android-components git:(codeCovFix) ✗ ls components/browser/storage-sync/build/jacoco/
testDebugUnitTest.exec   testReleaseUnitTest.exec
```

So, at least that means we'll have execution data available for the `jacocoTestReport` task.

**Comment by grigoryk** [comment]
And the report is now being generated locally:
```
➜  android-components git:(codeCovFix) ✗ ls -Al components/browser/storage-sync/build/reports/jacoco/jacocoTestReport/html/index.html
-rw-r--r--  1 grishakruglov  staff  3066 12 Jan 16:11 components/browser/storage-sync/build/reports/jacoco/jacocoTestReport/html/index.html
```

**Comment by grigoryk** [comment]
Ouch.. So in Taskcluster, here's what we end up running (this is for `feature-accounts-push`):
```
[task 2021-01-13T05:02:00.033Z] executing ['bash', '-cx', 'automation/taskcluster/helper/get-secret.py -s project/mobile/android-components/public-tokens -k codecov -f .cc_token && ./gradlew -PgoogleRepo=file:///builds/worker/fetches/android-gradle-dependencies/google -PjcenterRepo=file:///builds/worker/fetches/android-gradle-dependencies/jcenter -PcentralRepo=file:///builds/worker/fetches/android-gradle-dependencies/central listRepositories -Pcoverage :feature-accounts-push:assemble :feature-accounts-push:assembleAndroidTest :feature-accounts-push:testRelease :feature-accounts-push:lintRelease githubBuildDetails && automation/taskcluster/action/upload_coverage_report.sh']
```

Notice how it's running `testRelease`. This will produce `jacoco/testReleaseUnitTest.exec` file, while in `build.gradle` execution data for `jacocoTestReport` is set as `'jacoco/testDebugUnitTest.exec'`... so, of course the task is skipped - it's looking for a non-existent `exec` file.

**Comment by grigoryk** [comment]
Okay, so in https://github.com/mozilla-mobile/android-components/pull/9385 it's upload reports. Great! I'm not 100% it's all correct now, since the reports as codecov processed them look strange. Hopefully it's just because it was trying to diff stuff against a 7mo base commit.

**Comment by pocmo** [comment]
Shit, I think I know why the coverage is low: A while ago RelEng created an optimization step that will try to figure out what changed and only build related components. That means that now we only get the coverage reports for those. :(

**Comment by grigoryk** [comment]
As discussed on Slack, the run in https://github.com/mozilla-mobile/android-components/pull/9385 should produce a full coverage report in theory, since `build.gradle` was touched and everything's executed as a result.

Looking at the logs (browser-icons, in this case), I see this:
```
[task 2021-01-13T07:22:11.193Z] [ant:jacocoReport] Classes in bundle 'browser-icons' do not match with execution data. For report generation the same class files must be used as at runtime.
[task 2021-01-13T07:22:11.193Z] [ant:jacocoReport] Execution data for class mozilla/components/browser/icons/loader/HttpIconLoader does not match.
[task 2021-01-13T07:22:11.193Z] [ant:jacocoReport] Execution data for class mozilla/components/browser/icons/loader/IconLoader$Result$NoResult does not match.
[task 2021-01-13T07:22:11.193Z] [ant:jacocoReport] Execution data for class mozilla/components/browser/icons/loader/FailureCache does not match.
[task 2021-01-13T07:22:11.193Z] [ant:jacocoReport] Execution data for class mozilla/components/browser/icons/loader/MemoryIconLoader does not match.
[task 2021-01-13T07:22:11.193Z] [ant:jacocoReport] Execution data for class mozilla/components/browser/icons/utils/UtilsKt does not match.
...
```

Perhaps that's a clue.

**Comment by grigoryk** [comment]
I've added a commit to retain a generated jacoco coverage html report in taskcluster, and re-ran the CI. Here's an example: https://firefoxci.taskcluster-artifacts.net/A0Z0yGaISlyWfICFfowkxg/0/public/reports/jacoco/jacocoTestReport/html/index.html - so, that matches with what codecov is reporting (almost no coverage when there should be).

**Comment by grigoryk** [comment]
Okay, there are two problems:
1) we're hardcoding `kotlin-classes` and `intermediates/classes` paths to use debug build variants (and not the `release` ones, hence the classid mismatch)
2) overall we're producing an incorrect dependency graph - task `jacocoTestReport` that's hard-coded against a single build variant is set to finalize test tasks for _every_ variant

Fixing these definitions to run within a `android.buildTypes.all { buildType -> ... }` block and setting the dynamic `buildType.name` correctly in the task config gets what looks like correct coverage reports locally!

**Comment by grigoryk** [comment]
So, the potential remaining problem is what @pocmo mentioned - we no longer run every test on a PR push, just those that are relevant. Which means that unless codecov is smart enough about this (i'm trying to see if it is..), our coverage diffs in PRs are going to be incorrect. Well, at the very least we still get a baseline coverage for the `master` branch.

**Comment by grigoryk** [comment]
Okay, this landed in #9385 and we're running coverage reports for commits landing on `master` and for PRs. PR "% of diff covered" is working well. What's not working currently is that when stuff lands on `master`, we're not running every test task (as I assumed we will be), just the affected ones, so the reports are all over the place. Next step is to disable that optimization for `master` pushes.
